driver-embedded/src/main/com/mongodb/embedded/client/MongoDBCAPIHelper.java

```diff
@@ -0,0 +1,171 @@
+import com.sun.jna.NativeLibrary;
+import com.sun.jna.Pointer;
+import com.sun.jna.ptr.IntByReference;
+import com.sun.jna.ptr.PointerByReference;
+
+import static java.lang.String.format;
+
+final class MongoDBCAPIHelper {
+    private static final String NATIVE_LIBRARY_NAME = "mongo_embedded_capi";
+    private static MongoDBCAPI mongoDBCAPI;
+
+    static synchronized void checkHasBeenInitialized() {
+        checkInitialized();
+    }
+
+    // CHECKSTYLE.OFF: MethodName
+
+    static synchronized void init(final MongoEmbeddedSettings mongoEmbeddedSettings, final String config) {
+        if (mongoDBCAPI != null) {
+            throw new MongoClientException("MongoDBCAPI has been initialized but not closed");
+        }
+
+        if (mongoEmbeddedSettings.getLibraryPath() != null) {
+            NativeLibrary.addSearchPath(NATIVE_LIBRARY_NAME, mongoEmbeddedSettings.getLibraryPath());
+        }
+
+        try {
+            mongoDBCAPI = Native.loadLibrary(NATIVE_LIBRARY_NAME, MongoDBCAPI.class);
+        } catch (UnsatisfiedLinkError e) {
+            throw new MongoClientException(format("Failed to load the mongodb library: '%s'."
+                    + "%n %s %n"
+                    + "%n Please set the library location by either:"
+                    + "%n - Adding it to the classpath."
+                    + "%n - Setting 'jna.library.path' system property"
+                    + "%n - Configuring it in the 'MongoEmbeddedSettings.builder().libraryPath' method."
+                    + "%n", NATIVE_LIBRARY_NAME, e.getMessage()), e);
+        }
+        try {
+            validateErrorCode(mongoDBCAPI.libmongodbcapi_init(config));
+        } catch (Throwable t) {
+            throw createError("init", t);
+        }
+    }
+
+    static synchronized void fini() {
+        checkInitialized();
+        try {
+            validateErrorCode(mongoDBCAPI.libmongodbcapi_fini());
+            mongoDBCAPI = null;
+        } catch (Throwable t) {
+            throw createError("fini", t);
+        }
+    }
+
+    static synchronized Pointer db_new(final int argc, final String[] argv, final String[] envp) {
+        checkInitialized();
+        try {
+            Pointer val = mongoDBCAPI.libmongodbcapi_db_new(argc, argv, envp);
+            return val;
+        } catch (Throwable t) {
+            throw createError("db_new", t);
+        }
+    }
+
+    static synchronized void db_destroy(final Pointer db) {
+        checkInitialized();
+        try {
+            mongoDBCAPI.libmongodbcapi_db_destroy(db);
+        } catch (Throwable t) {
+            throw createError("db_destroy", t);
+        }
+    }
+
+    static synchronized void db_pump(final Pointer db) {
+        checkInitialized();
+        try {
+            validateErrorCode(mongoDBCAPI.libmongodbcapi_db_pump(db));
+        } catch (Throwable t) {
+            throw createError("db_pump", t);
+        }
+    }
+
+    static synchronized Pointer db_client_new(final Pointer db) {
```

Comment:
If we synchronize all these methods, especially db_client_wire_protocol_rpc, we've effectively created a single-threaded driver, and that's not what we want, as the embedded server is multi-thread safe.

The only synchronization that is required is the access to the mongoDBCAPI global, rather  than the entire method bodies.  That can be accomplished by making mongoDBCAPI volatile.  

It's ok to leave init and fini as synchronized, in order to avoid concurrent access to those blocks, but other than that I think the volatile is enough.

Reply:
Done